AWS: Use executor service by default when performing batch deletion of files

amogh-jahagirdar opened the pull request:
Update the S3FileIO to lazily load batches and use the existing deletion threadpool for performing concurrent S3#RemoveObjects calls. 

This will be used in subsequent PRs for performing bulk deletes in procedures like removing orphan files, snapshot expiration and purging the data, manifests, old metadata files during table drop.

amogh-jahagirdar commented:
I'm running AWS integ tests to validate this.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/io/SupportsBulkOperations.java:
```diff
@@ -18,12 +18,23 @@
+   * @param pathsToDelete The paths to delete
+   * @throws BulkDeletionFailureException in case of failure to delete at least 1 file
+   */
+  void deleteFiles(Iterable<String> pathsToDelete, ExecutorService svc)
```
Are we sure passing an executor service to `FileIO` makes sense?
Also, this logic seems reasonable for regular non-bulk IO implementations.

jackye1995 replied:
yeah, can we just use `ThreadPools.getWorkerPool`?

amogh-jahagirdar replied:
Good point, I missed there's already a threadpool in the fileIO implementation. For now, since it's unclear if passing it in the mixin interface makes sense (I thought it did because I was thinking a compute layer at a higher layer can determine what is the appropriate executor service to use and then pass that to the file io layer for better control, but maybe it doesn't matter). 

I've updated to us the existing executor service in S3FileIO.

jackye1995 commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -167,6 +168,29 @@ public Map<String, String> properties() {
+    Iterable<List<String>> pathBatches =
+        Iterables.partition(paths, awsProperties.s3FileIoDeleteBatchSize());
+    Tasks.foreach(pathBatches)
+        .noRetry()
```
can we retry during 503 throttling?

amogh-jahagirdar replied:
Missed this, but I think the default S3 SDK policy is to retry on throttling? I will double check that. 

would we want to add another layer of retries considering users can pass in their own S3 client with perhaps their own retry policy?

amogh-jahagirdar replied:
https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/retry/RetryMode.html#defaultRetryMode--

So there are a few locations and an environment variable which are attempted to be read for determining the default retry mode before defaulting to the "LEGACY" setting which is the case we would hit for our default S3 client provided in the default aws client factory. The legacy setting performs 3 retries; the docs mention that no token is subtracted from the token bucket when a request gets throttled but considering the max 3 retries and the fact that throttles are retried.

Considering that the s3 client is pluggable at runtime, users can also tune according to their desire. Although maybe for max retries, and policies we could also have aws properties to avoid the user having to create the s3 client themselves and pass it in, which we could tackle separately?

Let me know your thoughts and if I'm missing something!

jackye1995 replied:
yes I agree, then this looks good to me

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -168,56 +169,42 @@ public Map<String, String> properties() {
-    if (!awsProperties.isS3DeleteEnabled()) {
-      return;
+    if (awsProperties.isS3DeleteEnabled()) {
+      SetMultimap<String, String> bucketToObjects = computeBucketToObjects(paths);
```
We are now eagerly computing the bucket to objects mapping up front, as opposed to before where we would iterate over the paths, keep track of the objects per bucket and if for a given bucket the size of the objects is the batch size, the deletion would get triggered (and the mapping would get removed). 

Now, it's all up front, so there would be more memory consumption but I think it should be ok. 1 million objects with a max key of 1024 bytes is 1 GB representation of paths.

szehon-ho replied:
1gb does sounds like a lot , example on the spark driver.  Why cant we do it the old way?

jackye1995 replied:
can we try to use https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/paginators/ListObjectsV2Iterable.html so that the list is dynamically loaded instead of buffered upfront

amogh-jahagirdar replied:
@jackye1995 I don't think we need a separate API actually, we are already given the list of paths. We can still do it lazily by basically using the previous approach and constructing the bucket/key from the path and keeping track of when the objects for a bucket hits a certain batch size. It's just that using the Tasks framework in its current state becomes infeasible since it operates on a list and also will wait for completion. We want to just submit the batch deletion and move on, and then at the very end check the statuses of all those tasks. So using the underlying executor service fits that pattern. Let me know what you think.

jackye1995 replied:
Sounds good!

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -244,26 +231,34 @@ private void tagFileToDelete(String path, Set<Tag> deleteTags) throws S3Exceptio
+  private List<String> batchDeleteObjects(
+      String bucket, Iterable<String> objects, List<String> failures)
+      throws BulkDeletionFailureException {
+    Iterable<List<String>> batches =
```
Now that we're parallelizing by default we need to partition up front as well so that we can clearly assign a partition to a thread.

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -168,56 +169,42 @@ public Map<String, String> properties() {
+      bucketToObjects
+          .asMap()
+          .entrySet()
+          .forEach(
```
Technically, we could get even more throughput if the data is spread over multiple buckets, and parallelize this. But it's important that we do not parallelize over the buckets using the deletion threadpool otherwise we could easily deadlock. Another pool would need to be defined/used.

amogh-jahagirdar commented on aws/src/test/java/org/apache/iceberg/aws/s3/TestS3FileIO.java:
```diff
@@ -81,7 +81,9 @@ public class TestS3FileIO {
-          Integer.toString(batchDeletionSize));
+          Integer.toString(batchDeletionSize),
+          "s3.delete.num-threads",
+          "1");
```
S3Mock is not handling concurrent deletions well and now that multi-threading batch delete is passing, the testPrefixDelete is failing.. Our integ tests should give us good coverage in this regard.

amogh-jahagirdar commented on aws/src/integration/java/org/apache/iceberg/aws/s3/TestS3FileIOIntegration.java:
```diff
@@ -338,10 +338,10 @@ public void testDeleteFilesSingleBatchWithRemainder() throws Exception {
     S3FileIO s3FileIO = new S3FileIO(clientFactory::s3);
     List<Integer> scaleSizes = Lists.newArrayList(1, 1000, 2500);
-    String listPrefix = String.format("s3://%s/%s", bucketName, "prefix-list-test");
+    String listPrefix = String.format("s3://%s/%s/%s", bucketName, prefix, "prefix-list-test");
```
The cleanup of the AWS integ tests involves cleaning up under bucket/prefix where prefix is a UUID for the run of this test class. So the prefix list test actually should go under this as well otherwise multiple runs of the integ test for the same bcket will fail because there will be more objects than expected.

szehon-ho replied:
Is this related to this pr, or to fix another issue?  Should it be split out to not confuse this pr?

amogh-jahagirdar replied:
It's not directly related to this PR but I encountered failures in integ tests when validating this change and the fix seemed simple enough. I can raise it separately if it makes it easier to review! @szehon-ho

szehon-ho replied:
Yea that would be great for me !

amogh-jahagirdar replied:
Cool, I split the integ test changes into this PR: https://github.com/apache/iceberg/pull/5413

I will remove those changes from this branch, and when the other PR gets merged rebase off that

szehon-ho replied:
That one is merged now.

amogh-jahagirdar replied:
Thanks for the review on that @szehon-ho !

szehon-ho reviewed (commented):
I think this is probably a valid change, but as of now it doesnt seem to match the pr description ?  The pr description is about using an executor service.  Can you clarify the pr description?

amogh-jahagirdar commented:
> I think this is probably a valid change, but as of now it doesnt seem to match the pr description ? The pr description is about using an executor service. Can you clarify the pr description?

Sure, what I meant is we're now using the existing deletion threadpool (executor service) when performing batch deletes. Let me clarify the description. I also removed the integ test fixes, and moved them here https://github.com/apache/iceberg/pull/5413 . Thanks!

amogh-jahagirdar commented on aws/src/integration/java/org/apache/iceberg/aws/s3/TestS3FileIOIntegration.java:
```diff
@@ -376,8 +376,8 @@ private void testDeleteFiles(int numObjects, S3FileIO s3FileIO) throws Exception
       String deletionKey = objectKey + "-deletion-" + i;
-      write(s3FileIO, String.format("s3://%s/%s", bucketName, deletionKey));
-      paths.add(String.format("s3://%s/%s", bucketName, deletionKey));
+      write(s3FileIO, String.format("s3://%s/%s/%s", bucketName, prefix, deletionKey));
+      paths.add(String.format("s3://%s/%s/%s", bucketName, prefix, deletionKey));
```
This integ test fix is related to this PR since it is testing the batch deletion path. This is the same fix done in https://github.com/apache/iceberg/pull/5383 for adding the test prefix in the path , so that we are always guaranteed at the end of the test to cleanup these created objects.

jackye1995 commented on aws/src/test/java/org/apache/iceberg/aws/s3/TestS3FileIO.java:
```diff
@@ -229,6 +230,7 @@ public void testPrefixList() {
   }
 
   @Test
+  @Ignore
```
can you add some comments explaining why it's ignored

amogh-jahagirdar replied:
Done!

jackye1995 commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -182,41 +184,55 @@ public void deleteFiles(Iterable<String> paths) throws BulkDeletionFailureExcept
-          failedPath -> LOG.warn("Failed to delete object at path {}", failedPath));
-      numberOfFailedDeletions += failedDeletions.size();
-    }
+      int numberFailedDeletions = 0;
```
nit: just `failedDeletions`

amogh-jahagirdar replied:
failedDeletions is already used for the task level failed deletions. I renamed the overall counter to totalFailedDeletions so it's more obvious, since it represents the total number of failures across all tasks.

aokolnychyi commented:
Let me take another look tomorrow. Sorry for the delay!

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -182,41 +184,55 @@ public void deleteFiles(Iterable<String> paths) throws BulkDeletionFailureExcept
+          LOG.warn("Caught unexpected exception during batch deletion: ", e.getCause());
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw new RuntimeException("Interrupted when waiting for deletions to complete", e);
```
That's a great catch. I think we should. If delete files gets interrupted for any reason it makes sense to interrupt the ongoing work because there's no point in continuing them and utilizing any more resources. I will update for this

aokolnychyi replied:
This is why I usually recommend using `Tasks` for any concurrent code. The current implementation seems reasonable and starts execution as soon as there is a batch so we can keep it.

Just as a reference, one could use `BinPacking` and `Tasks` for this type of work in the future .

amogh-jahagirdar replied:
Thanks for the pointer @aokolnychyi , this is a pattern I will take a look at and clean it up in a subsequent PR if it applies!

aokolnychyi reviewed (approved):
This change looks good to me.

After a closer look, passing an explicit executor started to make more sense to me. I think we may add that overloaded method back once we consume these changes in other places.

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -182,41 +184,55 @@ public void deleteFiles(Iterable<String> paths) throws BulkDeletionFailureExcept
+        try {
+          List<String> failedDeletions = deletionTask.get();
+          failedDeletions.forEach(
+              failedPath -> LOG.warn("Failed to delete object at path {}", failedPath));
```
Updated!

amogh-jahagirdar commented:
> This change looks good to me.
> 
> After a closer look, passing an explicit executor started to make more sense to me. I think we may add that overloaded method back once we consume these changes in other places.

Right, this is my thought as well. Once we see there is a need for the API we can add that, it's harder to go the other way.

Thanks for the review @aokolnychyi !

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -182,41 +184,55 @@ public void deleteFiles(Iterable<String> paths) throws BulkDeletionFailureExcept
+                .submit(
+                    () ->
+                        deleteBatch(
+                            bucketToObjectsEntry.getKey(), bucketToObjectsEntry.getValue()));
```
Updated!

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:
```diff
@@ -244,26 +260,33 @@ private void tagFileToDelete(String path, Set<Tag> deleteTags) throws S3Exceptio
+                .map(error -> String.format("s3://%s/%s", request.bucket(), error.key()))
+                .collect(Collectors.toList()));
       }
+    } catch (Exception e) {
```
Yeah I think in case of any failure we should surface a BulkDeletionFailure at the end. So catching the generic exception allows us to handle any failure, treat it as a failure to delete the entire batch, and add that to the failed files list. I'm not sure of any other case where we want to surface something else. We're logging the specific exception so that folks can debug.

jackye1995 commented:
I think we get enough approvals, thanks for the work @amogh-jahagirdar , and thanks everyone for the review!

amogh-jahagirdar commented:
Thanks everyone for the reviews!
